x/src/lint/whitespace.rs

```diff
@@ -1,55 +1,94 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-use std::{borrow::Cow, convert, ffi::OsStr, path::Path, str};
+use crate::lint::engine::prelude::*;
 
-pub(super) fn has_newline_at_eof(file: &Path, contents: &str) -> Result<(), Cow<'static, str>> {
-    if skip_whitespace_checks(file) {
-        return Ok(());
-    }
+#[derive(Clone, Copy, Debug)]
+pub(super) struct EofNewline;
 
-    if !contents.ends_with('\n') {
-        Err("missing a newline at EOF".into())
-    } else {
-        Ok(())
+impl Linter for EofNewline {
+    fn name(&self) -> &'static str {
+        "eof-newline"
     }
 }
 
-pub(super) fn has_trailing_whitespace(
-    file: &Path,
-    contents: &str,
-) -> Result<(), Cow<'static, str>> {
-    if skip_whitespace_checks(file) {
-        return Ok(());
+impl ContentLinter for EofNewline {
+    fn pre_run<'l>(&self, file_ctx: &FileContext<'l>) -> Result<RunStatus<'l>> {
+        Ok(skip_whitespace_checks(file_ctx.extension()))
     }
 
-    for (ln, line) in contents
-        .lines()
-        .enumerate()
-        .map(|(ln, line)| (ln + 1, line))
-    {
-        if line.trim_end() != line {
-            return Err(Cow::Owned(format!("trailing whitespace on line {}", ln)));
+    fn run<'l>(
+        &self,
+        ctx: &ContentContext<'l>,
+        out: &mut LintFormatter<'l, '_>,
+    ) -> Result<RunStatus<'l>> {
+        let content = match ctx.content() {
+            Some(text) => text,
+            None => return Ok(RunStatus::Skipped(SkipReason::NonUtf8)),
+        };
+        if !content.ends_with('\n') {
+            out.write(LintMessage::new(LintLevel::Error, "missing newline at EOF"));
         }
+        Ok(RunStatus::Executed)
     }
+}
 
-    if contents
-        .lines()
-        .rev()
-        .take_while(|line| line.is_empty())
-        .count()
-        > 0
-    {
-        return Err("trailing whitespace at EOF".into());
+#[derive(Clone, Copy, Debug)]
+pub(super) struct TrailingWhitespace;
+
+impl Linter for TrailingWhitespace {
+    fn name(&self) -> &'static str {
+        "trailing-whitespace"
     }
+}
+
+impl ContentLinter for TrailingWhitespace {
+    fn pre_run<'l>(&self, file_ctx: &FileContext<'l>) -> Result<RunStatus<'l>> {
+        Ok(skip_whitespace_checks(file_ctx.extension()))
+    }
+
+    fn run<'l>(
+        &self,
+        ctx: &ContentContext<'l>,
+        out: &mut LintFormatter<'l, '_>,
+    ) -> Result<RunStatus<'l>> {
+        let content = match ctx.content() {
+            Some(text) => text,
+            None => return Ok(RunStatus::Skipped(SkipReason::NonUtf8)),
+        };
+        if !content.ends_with('\n') {
+            out.write(LintMessage::new(LintLevel::Error, "missing newline at EOF"));
+        }
```

Comment:
Does this check need to be duplicated here? (since it is also in the lint above)

Reply:
nope, whoops, sorry!